vod/hls/m3u8_builder.c

```diff
@@ -810,6 +816,70 @@ m3u8_builder_append_index_url(
 	return p;
 }
 
+static size_t
+m3u8_builder_closed_captions_get_size(
+	media_set_t* media_set,
+	request_context_t* request_context)
+{
+	media_closed_captions_t* closed_captions;
+	size_t result = 0;
+	size_t base =
+		sizeof("\n\n") - 1 +
+		(sizeof(M3U8_EXT_MEDIA_BASE) - 1 + VOD_INT32_LEN +
+		sizeof(M3U8_EXT_MEDIA_TYPE_CLOSED_CAPTIONS) - 1 +
+		sizeof(M3U8_EXT_MEDIA_GROUP_ID_CLOSED_CAPTIONS) - 1 +
+		sizeof(M3U8_EXT_MEDIA_LANG) - 1 +
+		LANG_ISO639_3_LEN +
+		sizeof(M3U8_EXT_MEDIA_INSTREAM_ID) - 1 +
+		sizeof(M3U8_EXT_MEDIA_NON_DEFAULT) - 1 +
+		sizeof("\"\n") - 1);
+
+	for (closed_captions = media_set->closed_captions; closed_captions < media_set->closed_captions_end; closed_captions++)
+	{
+		result += base + closed_captions->id.len + closed_captions->label.len;
+	}
+
+	return result;
+}
+
+static u_char*
+m3u8_builder_closed_captions_write(
+	u_char* p,
+	media_set_t* media_set)
+{
+	media_closed_captions_t* closed_captions;
+	bool_t first_cc_track = TRUE;
+
+	for (closed_captions = media_set->closed_captions; closed_captions < media_set->closed_captions_end; closed_captions++)
+	{
+		p = vod_copy(p, M3U8_EXT_MEDIA_BASE_CC, sizeof(M3U8_EXT_MEDIA_BASE_CC) - 1);
+		
+		p = vod_sprintf(p, M3U8_EXT_MEDIA_NAME, (char *) closed_captions->label.data);
+
+		p = vod_sprintf(p, M3U8_EXT_MEDIA_LANG,
+				lang_get_rfc_5646_name(closed_captions->language));
+
+		if (first_cc_track)
```

Comment:
for consistency, better remove this bool and use closed_captions == media_set->closed_captions